### PR TITLE
🐛 Watch external objects for machine before deleting

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:
This fixes a race condition in the machine controller when the controller is restarted after a machine has been marked for deletion but before the infra machine or bootstrap config are deleted. In that case, the controller doesn't have watches on those external objects, so the reconciliation is not triggered once they are deleted. This means that Machines stay in Deleting state for the resync period, the default is 10 minutes.

/area machine

### CODE REVIEW

File: internal/controllers/machine/machine_controller_phases.go [hunk 1]
```diff
@@ -89,12 +87,44 @@ func (r *Reconciler) reconcilePhase(_ context.Context, m *clusterv1.Machine) {
...
+		m.Status.FailureReason = &machineStatusError
+	}
+	if failureMessage != "" {
+		m.Status.FailureMessage = ptr.To(
+			fmt.Sprintf("Failure detected from referenced resource %v with name %q: %s",
+				obj.GroupVersionKind(), obj.GetName(), failureMessage),
+		)
+	}
+
+	return external.ReconcileOutput{Result: obj}, nil
```
Comment (sbueringer) [inline internal/controllers/machine/machine_controller_phases.go:120]: This code is now run twice. I assume it shouldn't be also part of ensureExternalOwnershipAndWatch

**Review by sbueringer — COMMENTED** [review]
Very nice! Just one finding

File: internal/controllers/machine/machine_controller_phases.go (hunk 1 above)
Comment (g-gaston) [reply internal/controllers/machine/machine_controller_phases.go:120]: you are 100% correct, I'll remove it from there
good catch, thanks!

**Comment by sbueringer** [comment]
Thx!

/lgtm

/assign @fabriziopandini @vincepri @chrischdi 

I think we should backport

**Review by chrischdi — COMMENTED** [review]
/approve

**Pull request merged**